### PR TITLE
Bug #2845 typescript angular inheritance

### LINKED ISSUE #2845: [BUG] Inheritance is broken
#### Bug Report Checklist

- [x] Have you provided a full/minimal spec to reproduce the issue?
- [x] Have you validated the input using an OpenAPI validator ([example](https://apidevtools.org/swagger-parser/online/))?
- [x] What's the version of OpenAPI Generator used?
- [x] Have you search for related issues/PRs?
- [x] What's the actual output vs expected output?

##### Description

Basic inheritance, [as described in the OpenAPI spec](https://swagger.io/docs/specification/data-models/inheritance-and-polymorphism/) does not work.

Instead of creating a class / interface hierarchy using inheritance to represent the composition, all properties are flattened into the subclass.

##### openapi-generator version

[4.0.0-20190508.072036-627](https://oss.sonatype.org/content/repositories/snapshots/org/openapitools/openapi-generator-cli/4.0.0-SNAPSHOT/openapi-generator-cli-4.0.0-20190508.072036-627.jar) or master

##### OpenAPI declaration file content or url

schema.yaml:

```yaml
openapi: "3.0.1"
info:
  version: 1.0.0
  title: Inheritance test
paths:
  /order:
    get:
      summary: Product order details
      operationId: getOrder
      responses:
        '500':
          description: Successful load
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/ExtendedErrorModel'
components:
  schemas:
    BasicErrorModel:
      type: object
      required:
        - message
        - code
      properties:
        message:
          type: string
        code:
          type: integer
          minimum: 100
          maximum: 600
    ExtendedErrorModel:
      allOf:
        - $ref: '#/components/schemas/BasicErrorModel'
        - type: object
          required:
            - rootCause
          properties:
            rootCause:
              type: string

```

##### Expected output

The following class/interface hierarchy: to be generated:

- `BasicErrorModel` <- `ExtendedErrorModel`

##### Actual output

Three unrelated classes/interfaces:

 - `BasicErrorModel`
 - `ExtendedErrorModel`
 - `ExtendedErrorModelAllOf`

##### Command line used for generation

```
java -jar openapi-generator-cli.jar generate -i ./schema.yaml -g typescript-angular -o src
```

I have also tried with the Java generator, and it creates the same hierarchy, so it is not just a Typescript problem.

##### Steps to reproduce

1. Copy the above yaml into `schema.yaml` locally.
2. Run the command above.
3. Examine the generated code.

##### Suggest a fix

I'm not familiar with the codebase, but it appears that `getParentName` in `ModelUtils.java` only considers a class as a viable parent for inheritance if it has a discriminator, but the above inheritance pattern does not use discriminators.

### PR DESCRIPTION
**Pull request by mnahkies**
### PR checklist

- [X] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [X] Ran the shell script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh`, `./bin/openapi3/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates). Windows batch files can be found in `.\bin\windows\`. If contributing template-only or documentation-only changes which will change sample output, be sure to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) first.
- [X] Filed the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `4.1.x`, `5.0.x`. Default: `master`.
- [X] Copied the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

@TiFu (2017/07) @taxpon (2017/07) @sebastianhaas (2017/07) @kenisteward (2017/07) @Vrolijkx (2017/09) @macjohnny (2018/01) @nicokoenig (2018/09) @topce (2018/10) @akehir (2019/07)
### Description of the PR

It appears that the `typescript-angular` template was re-written to expect the `supportsMultipleInheritance` flag to be enabled in 8c599ebf12483495f889cfc07ba6988d68b5e3ca but that the flag wasn't actually enabled.

By enabling this flag, parents are correctly added to interfaces as expected.

Also ran `./bin/openapi3/typescript-angular-petstore-all.sh` before and after making changes. Seems this hadn't been run on master so there is some noise from that. Unfortunately the sample yaml used doesn't seem to exercise the allOf, anyOf, oneOf keywords and so no change occurred after the fix.

Fixes #2845

### CODE REVIEW

**Comment by macjohnny** [comment]
Please re-generate the samples with the 
./bin/typescript-angular-petstore-all.sh
Command only

**Comment by macjohnny** [comment]
Should the flag be added for typescript-fetch, too?

File: samples/client/petstore/typescript-angular-v2/default/api/pet.service.ts [hunk 1]
```diff
@@ -63,10 +63,10 @@ export class PetService {
     /**
      * 
      * @summary Add a new pet to the store
-     * @param body Pet object that needs to be added to the store
+     * @param pet Pet object that needs to be added to the store
      */
-    public addPet(body: Pet, extraHttpRequestParams?: RequestOptionsArgs): Observable<{}> {
-        return this.addPetWithHttpInfo(body, extraHttpRequestParams)
+    public addPet(pet: Pet, extraHttpRequestParams?: RequestOptionsArgs): Observable<{}> {
```
Comment (wing328) [inline samples/client/petstore/typescript-angular-v2/default/api/pet.service.ts]: @mnahkies thanks for the PR. Looks like you're using an old version of OpenAPI Generator as the body name in OAS v2 is not preserved. Can you please pull the latest master into your branch and regenerate the samples again?

**Comment by wing328** [comment]
> Unfortunately the sample yaml used doesn't seem to exercise the allOf, anyOf, oneOf keywords and so no change occurred after the fix.

We've some test spec (OAS v3) for allOf, anyOf, oneOf under https://github.com/OpenAPITools/openapi-generator/tree/master/modules/openapi-generator/src/test/resources/3_0

File: samples/client/petstore/typescript-angular-v2/default/api/pet.service.ts (hunk 1 above)
Comment (mnahkies) [reply samples/client/petstore/typescript-angular-v2/default/api/pet.service.ts]: Hmm, seems like this script actually generates to the same directory as the openapi2 script, is that intentional?
```
$ mvn clean package && ./bin/openapi3/typescript-angular-petstore-all.sh
```

Results in no changes (based on latest master)

Running 
```
$ mvn clean package && ./bin/typescript-angular-petstore-all.sh
```

Reverts all changes to samples to current master. I'll push this, but seems like some work needs to be done around the typescript client samples with respect to openapi3

**Comment by wing328** [comment]
@mnahkies do you have a spec to reproduce the error? sorry if I miss it in your previous conversation. Just want to perform a bit more tests before merging this PR.

Thanks again for the PR.

**Comment by mnahkies** [comment]
@wing328 this is the spec I was testing with. Also I missed deleting some files that were generating when I ran the openapi3 script but shouldn't exist after running the openapi2 script.

```
openapi: "3.0.1"
info:
  version: 1.0.0
  title: Inheritance test
paths:
  /pets:
    patch:
      requestBody:
        content:
          application/json:
            schema:
              oneOf:
                - $ref: '#/components/schemas/Cat'
                - $ref: '#/components/schemas/Dog'
              discriminator:
                propertyName: pet_type
      responses:
        '200':
          description: Updated
components:
  schemas:
    Pet:
      type: object
      required:
        - pet_type
      properties:
        pet_type:
          type: string
      discriminator:
        propertyName: pet_type
    Dog:     # "Dog" is a value for the pet_type property (the discriminator value)
      allOf: # Combines the main `Pet` schema with `Dog`-specific properties 
        - $ref: '#/components/schemas/Pet'
        - type: object
          # all other properties specific to a `Dog`
          properties:
            bark:
              type: boolean
            breed:
              type: string
              enum: [Dingo, Husky, Retriever, Shepherd]
    Cat:     # "Cat" is a value for the pet_type property (the discriminator value)
      allOf: # Combines the main `Pet` schema with `Cat`-specific properties 
        - $ref: '#/components/schemas/Pet'
        - type: object
          # all other properties specific to a `Cat`
          properties:
            hunts:
              type: boolean
            age:
              type: integer
```
[example.txt](https://github.com/OpenAPITools/openapi-generator/files/3569930/example.txt)

**Comment by wing328** [comment]
Did some tests and notice the following changes when compared with the master:
```
 import { Pet } from './pet';
 
-export interface Cat { 
+export interface Cat extends Pet { 
     hunts?: boolean;
     age?: number;
 }
```

Let's go with this fix and I'll merge it once the CircleCI tests pass.

**Comment by mnahkies** [comment]
Thanks @wing328 - will this release also go to maven? https://mvnrepository.com/artifact/org.openapitools/openapi-generator-maven-plugin

**Comment by wing328** [comment]
https://search.maven.org/artifact/org.openapitools/openapi-generator-maven-plugin/4.1.2/maven-plugin

The link you provided seems not up-to-date.

**Pull request merged**